src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs

```diff
@@ -166,39 +186,52 @@ public void Stop()
         public void SetEventHandler<TParams>(
             EventType<TParams> eventType,
             Func<TParams, EventContext, Task> eventHandler,
-            bool overrideExisting)
+            bool overrideExisting,
+            bool isParallelProcessingSupported = false)
         {
             if (overrideExisting)
             {
                 // Remove the existing handler so a new one can be set
                 this.eventHandlers.Remove(eventType.MethodName);
             }
 
+            this.eventHandlerParallelismMap.Add(eventType.MethodName, isParallelProcessingSupported);
             this.eventHandlers.Add(
                 eventType.MethodName,
                 async (eventMessage, messageWriter) =>
                 {
-                    var eventContext = new EventContext(messageWriter);
-                    TParams typedParams = default(TParams);
-                    try
-                    {                
-                        if (eventMessage.Contents != null)
-                        {
-                            try
-                            {
-                                typedParams = eventMessage.Contents.ToObject<TParams>();
-                            }
-                            catch (Exception ex)
+                    Func<Task> func = async () =>
```

Comment:
What you're defining here with `async (eventMessage, messageWriter) => { }` is already a function, why are you creating another function `Func<Task>` inside and awaiting on it?

Same comment for above method.

Reply:
Pasting a simple example for reference, no code would be executed until you await on the function, as this is an async task object you're passing around.

```
Func<int, Task> func = async(i) =>
{
    Console.WriteLine("2. Non awaited result");
    await Console.Out.WriteLineAsync("3. Awaited result");
};
var function = func;
await Task.Delay(100);
await Console.Out.WriteLineAsync("1. Before await");
await function(1);
```

Reply:
Also tested the original code from @alanrenmsft , it doesn't have negative consequences from execution perspective. 
It works normally as expected:

```
Func<int, Task> func = (i) => Task.Run(async () => 
{
    Console.WriteLine("2. Non awaited result");
    await Console.Out.WriteLineAsync("3. Awaited result");
});
var function = func;
await Task.Delay(100);
await Console.Out.WriteLineAsync("1. Before await");
await function(1);
```

Reply:
The difference here is that the function actually gets invoked here with `handlerToAwait = requestHandler(messageToDispatch, messageWriter);`, I'll move this call to where await happens.